lib/puppet/provider/foreman_resource/rest_v3.rb

```diff
@@ -96,6 +99,20 @@ def success?(response)
   end
 
   def error_message(response)
-    JSON.parse(response.body)['error']['full_messages'].join(' ') rescue "unknown error (response #{response.code})"
+    explanations = {
+      '400' => 'Something is wrong with the data sent to Foreman server',
```

Comment:
Thinking out loud because it's a bit vague. How about
```suggestion
      '400' => 'Receiving server (Foreman) did not accept our data',
```

The reason I put Foreman between braces is that if it actually hits some proxy or something else in between then it may be clearer. Of course there shouldn't be anything in between but if everything went well we wouldn't be writing this code :) I believe 400 is more likely to be returned by some non-Foreman server in between. Like if a user pointed it at `https://internal-service.example.com`

Reply:
Thanks. I'm not fully satisfied with that wording either (and I'm not sure that I'll ever be).

Thinking longer term, I think the best approach is to have links in these explanation messages to community wikis that can be much more comprehensive in covering various cases, with the option of downstream releases replacing those links with more specifically applicable ones for their community. Often the workflow for an user encountering one of these errors will be to copy and paste it into a search engine, so we might as well remove that step for them and just send them directly to the relevant webpage where we can really cover all possible causes in order of likelihood.

I'd like to consider this as a parking lot item to revisit in the future.